### DISCUSSION TITLE
Open telemetry java agent instrumentation: missing manifest file

### DISCUSSION THREAD

**Discussion by hkdev27** [open-telemetry/opentelemetry-java-instrumentation#4828] [Q&A] [answered]
I have created a spring boot project jar and running the following command and it works well. Application starts successfully and sends telemetry data to jaeger. 

java -javaagent:opentelemetry-javaagent.jar -Dotel.resource.attributes=service.name=test-service -Dotel.traces.exporter=jaeger -Dotel.exporter.jaeger.endpoint=http://localhost:14250 -jar test-service.jar

But it throws exception when trying to run as image. 
Dockerfile is as follows:

Variant 1: 
FROM openjdk:17-jdk-alpine
ARG JAR_FILE=*.jar
COPY ${JAR_FILE} test-service.jar
CMD ["java","-javaagent:/src/main/resources/opentelemetry-javaagent.jar -Dotel.resource.attributes=service.name=test-service -Dotel.traces.exporter=jaeger -Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Variant 2: 
FROM openjdk:17-jdk-alpine
ARG JAR_FILE=*.jar
COPY ${JAR_FILE} test-service.jar
CMD ["java","-javaagent:opentelemetry-javaagent.jar -Dotel.resource.attributes=service.name=test-service -Dotel.traces.exporter=jaeger -Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Exception:
Error opening zip file or JAR manifest missing : /src/main/resources/opentelemetry-javaagent.jar -Dotel.resource.attributes
Error occurred during initialization of VM
agent library failed to init: instrument

How can I overcome this issue. Please help.

**Comment by iNikem** [comment, marked as answer]
You have to put every single command line parameter into quotes:
`"-javaagent:/src/main/resources/opentelemetry-javaagent.jar", "-Dotel.resource.attributes=service.name=test-service", "-Dotel.traces.exporter=jaeger", "-Dotel.exporter.jaeger.endpoint=http://localhost:14250"`

**Reply by hkdev27** [reply]
Thanks for pointing it out. I did the changes, it progressed bit further but still not done. 

Variant 1: 
FROM openjdk:17-jdk-alpine
ARG JAR_FILE=*.jar
COPY ${JAR_FILE} test-service.jar
CMD ["java","-javaagent:opentelemetry-javaagent-1.9.1.jar" "-Dotel.resource.attributes=service.name=test-service" "-Dotel.traces.exporter=jaeger" "-Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Variant 2: 
FROM openjdk:17-jdk-alpine
ARG JAR_FILE=*.jar
COPY ${JAR_FILE} test-service.jar
CMD ["java","-javaagent:/BOOT-INF/lib/opentelemetry-javaagent-1.9.1.jar" "-Dotel.resource.attributes=service.name=test-service" "-Dotel.traces.exporter=jaeger" "-Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Exception has changed: 
Variant 1: /bin/sh: [java,-javaagent:opentelemetry-javaagent-1.9.1.jar: not found
Variant 2: /bin/sh: [java,-javaagent:/BOOT-INF/lib/opentelemetry-javaagent-1.9.1.jar: not found

I have placed the agent jar at:  C:\source\test-service\src\main\resources
It seems to be location issue of some sort, Can you suggest the correction.

**Reply by mateuszrzeszutek** [reply]
Did you `COPY` or `ADD` the agent at all? It kind of looks like the agent jar is just not there.

**Reply by hkdev27** [reply]
I did not copy the agent jar explicitly. I added the jar in classpath. 
Do I need to ADD jar while creating docker image.

**Reply by mateuszrzeszutek** [reply]
Yes, try ADDing or COPYing the agent.
It should not be added to the application classpath - the javaagent is not your "normal" library, it needs to be in a separate jar so that you can pass path to that jar as `-javaagent`.

**Reply by hkdev27** [reply]
Variant 1:
FROM openjdk:17-jdk-alpine
COPY test-service.jar test-service.jar
CMD ["java","-Dotel.resource.attributes=service.name=test-service" "-Dotel.traces.exporter=jaeger" "-Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Exception: /bin/sh: [java,-Dotel.resource.attributes=service.name=test-service: not found

Variant 2:
FROM openjdk:17-jdk-alpine
COPY test-service.jar test-service.jar
COPY opentelemetry-javaagent.jar opentelemetry-javaagent.jar
CMD ["java","-javaagent:opentelemetry-javaagent.jar" "-Dotel.resource.attributes=service.name=test-service" "-Dotel.traces.exporter=jaeger" "-Dotel.exporter.jaeger.endpoint=http://localhost:14250","-jar","/test-service.jar"]

Exception: /bin/sh: [java,-javaagent:opentelemetry-javaagent.jar: not found

**Reply by hkdev27** [reply]
Is there any other way to add agent as pom dependency or maven plugin. So that it can directly pick it up from lib.

**Reply by mateuszrzeszutek** [reply]
You're missing commas in a couple of places after the second CMD arg, try
```dockerfile
FROM openjdk:17-jdk-alpine
COPY test-service.jar test-service.jar
COPY opentelemetry-javaagent.jar opentelemetry-javaagent.jar
CMD ["java", "-javaagent:opentelemetry-javaagent.jar", "-Dotel.resource.attributes=service.name=test-service", "-Dotel.traces.exporter=jaeger", "-Dotel.exporter.jaeger.endpoint=http://localhost:14250", "-jar", "/test-service.jar"]
```

**Reply by hkdev27** [reply]
Thank you guys.
